### PR TITLE
sorted imports using isort and enabled isort check on travis

### PR DESCRIPTION
**Pull request by voith**
This PR sorts all unsorted imports using [isort](http://isort.readthedocs.io/en/latest/). I have also enabled isort check on travis. For the ease of reviewing I have split the PR into two separate commits, one for sorting all the imports and another for enabling isort check on tavis.

There are certain imports that I had to skip from sorting in order to handle import errors and deprecation warnings. Let me know if there are any other imports that have to skipped or the ones that I've skipped are unnecessary.

### CODE REVIEW

**Comment by voith** [comment]
Hey @redapple, Please have a look at this PR. `Django` uses isort. I thought `Scrapy` might find it useful too.

**Comment by redapple** [comment]
Hi @voith , thanks for the PR.
This is a very big change.
I have no strong opinion on sorting import (I'm a bit sloppy myself regarding this), but other maintainers may have other views.
At least, maybe the "isort" tox env could be allowed to fail on Travis.

**Comment by voith** [comment]
Thanks @redapple for your feedback. I have allowed tox env "isort" to fail on travis.
I would like to get the opinion of other maintainers if they'd think this PR would be useful for scrapy.
ping @kmike @eliasdorneles @dangra

**Comment by eliasdorneles** [comment]
No strong opinion here, either.

File: scrapy/downloadermiddlewares/httpcache.py [hunk 1]
```diff
@@ -1,10 +1,13 @@
 from email.utils import formatdate
+
 from twisted.internet import defer
-from twisted.internet.error import TimeoutError, DNSLookupError, \
-        ConnectionRefusedError, ConnectionDone, ConnectError, \
-        ConnectionLost, TCPTimedOutError
+from twisted.internet.error import (
+    ConnectError, ConnectionDone, ConnectionLost, ConnectionRefusedError, DNSLookupError, TCPTimedOutError,
```
Comment (kmike) [inline scrapy/downloadermiddlewares/httpcache.py]: line looks too long

File: scrapy/downloadermiddlewares/httpproxy.py [hunk 2]
```diff
@@ -1,15 +1,17 @@
...
+
+from scrapy.exceptions import NotConfigured
+from scrapy.utils.httpobj import urlparse_cached
+from scrapy.utils.python import to_bytes
+
 try:
     from urllib2 import _parse_proxy
 except ImportError:
     from urllib.request import _parse_proxy
-from six.moves.urllib.parse import urlunparse
```
Comment (kmike) [inline scrapy/downloadermiddlewares/httpproxy.py]: it is nice to have all urllib imports grouped

**Comment by kmike** [comment]
I haven't reviewed all changes, but overall impression is the following:

1. most of changes are good;
2. I like to think about `six` as of an stdlib import, because usually you import stdlib functions from it, but that's a personal preference;
3. try/except imports should be grouped with other imports of the same type imho;
4. some changes broke line length limit;
5. I'm not comfortable with requiring alphabetical order for all imports. One-time change is fine, but a CI failing because imports are not in alphabetical order is too much. I also sometimes prefer to order imports by their length, if they are stdlib imports :)
6. I usually also think about `__future__` imports as of stdlib imports (i.e. no extra blank line between future imports and stdlib imports), but that's a personal preference.

So +1 from me to do a one-time cleanup, making sure each change doesn't make code worse; -0 to put isort on CI.

**Comment by dangra** [comment]
I don't like forcing commits to pass the alphabetical sorting, that sounds arbitrary ridiculous to me. I prefer the length sort but even there I wouldn't make it required to pass CI.

Is there any styling guide about import sorting? I think PEP8 only mentions about different sections for stdlib, thirdparty and project imports. Different sections and no spurious imports are the only rules I would like CI to enforce if any.

**Comment by dangra** [comment]
I think linters like flake8, pylint, pep8, pyflakes... are much more useful tools to put in CI.

**Comment by voith** [comment]
I am going by @dangra's comment. Closing this

**Pull request closed without merging**